Add Typeahead Component

bameyrick opened the pull request:
### What is the context of this PR?
Added a typeahead component that should hopefully address the accessibility issues raised in the DAC report.

This should address:
* DAC_TECH_Forms_01
* DAC-USER-SRVO-T1-01
* DAC-USER-SRVO-T2-01
* DAC_TECH_Retest_Aria01

### How to review 
All new code is under `components/02-form-elements/typeahead`

Run `yarn start`

Standard country picker example (country list might not be 100% as they're generated from a library):
http://localhost:3000/components/detail/typeahead--country-typeahead

Welsh version of country picker (demonstrates alternative values functionality):
http://localhost:3000/components/detail/typeahead--country-typeahead-cy

* Check keyboard navigation
* Check VoiceOver
* Check label is transferred to input and is read on focus
* Check instructions make sense for touch device users also
* Check ability to select results after software keyboard is dismissed

armstrongb commented:
@bameyrick can you put in a list of items from the DAC report that you are remedying in this PR?

bameyrick commented:
@armstrongb No problem, I've updated the PR description now

armstrongb commented on components/02-form-elements/typeahead/typeahead.hbs:
```diff
@@ -0,0 +1,17 @@
+    {{> @label }}
+    <select class="typeahead__select js-typeahead-select input input--select{{#if select_classes}}{{select_classes}}{{/if}}" id="{{label_for}}" name="select">
+      {{#each options}}
+        <option{{#if option_value}} value="{{option_value}}"{{/if}}{{#if selected}} selected{{/if}}{{#if disabled}} disabled{{/if}}{{#if option_alternatives}}data-alternatives=" {{option_alternatives}}"{{/if}}>{{option_text}}</option>
```
- `value="GB"data-alternatives=" United`
should be:
- `value="GB" data-alternatives="United`

bameyrick replied:
This should be resolved now

armstrongb commented on components/02-form-elements/typeahead/typeahead.config.js:
```diff
@@ -0,0 +1,201 @@
+import { occupations } from './occupations';
+import { ethnicities } from './ethnicities';
+
+const territoriesJSON = {
```
**Query**
I thought that this was going to be using the .gov country register as this is the canonical source for gov work?
https://www.registers.service.gov.uk/registers/country

armstrongb replied:
cc @charlotte-hirst

bameyrick replied:
This is dummy data I've used because I wanted Welsh country names to test some functionality and I couldn't fine a Welsh version on the register. I wouldn't recommend using this data in production.

armstrongb commented on components/02-form-elements/typeahead/_typeahead.scss:
```diff
@@ -0,0 +1,85 @@
+
+    &--no-results {
+      cursor: not-allowed;
+      background: $color-dark-grey;
```
Contrast is `5.92:1` which passes at AA not AAA.

bameyrick replied:
I've lightened the grey which should hopefully improve the contrast ratio

armstrongb commented:
`French` & `Welsh` etc but `Spaniard`.

armstrongb commented:
**1.** If a value has been searched for or entered correctly e.g. `French` is the field is re-focussed then it displays the 'No results found' message. The input prefixes the value with a space and when the field is re-focussed the original value does not feature in the returned results.

**2.** If a _partial_ value has been used e.g. `Polynesian` the input pre-fixes the selected value with a space and when the field is re-focussed the original value does not feature in the returned results.
  1. Type 'Polynesian' returned results are `Polynesian, Niuean - Polynesian, Tokelauan - Polynesian island chain in Pacific Ocean`
  2. Select `Polynesian`
  3. Re-focus the typeahead and only `Niuean - Polynesian, Tokelauan - Polynesian island chain in Pacific Ocean` are available to select.
  4. delete the space from the start of the input and all three are shown.

armstrongb commented on components/02-form-elements/typeahead/typeahead.hbs:
```diff
@@ -0,0 +1,17 @@
+    {{> @label }}
+    <select class="typeahead__select js-typeahead-select input input--select{{#if select_classes}}{{select_classes}}{{/if}}" id="{{label_for}}" name="select">
+      {{#each options}}
+        <option{{#if option_value}} value="{{option_value}}"{{/if}}{{#if selected}} selected{{/if}}{{#if disabled}} disabled{{/if}}{{#if option_alternatives}}data-alternatives=" {{option_alternatives}}"{{/if}}>{{option_text}}</option>
```
### Ethnicity
`value="{{option_value}}"` appears to populate with spaces in the value. For example: `value=" Welsh"`

bameyrick replied:
This and the comment above about `French` and `Polynesian` should have been resolved from removing the leading spaces from the ethnicities data.

armstrongb commented:
`US` and `United States` appear to be different values.

and `United States of America` doesn't appear to be in the data.

bameyrick commented:
@will-madebymoon This should be resolved now

bameyrick commented:
@armstrongb So I've done a little digging into this; the aria-live announcements for result updates work fine with VoiceOver in Safari (see screenshot):
<img width="661" alt="screenshot 2018-10-18 at 13 25 44" src="https://user-images.githubusercontent.com/8447446/47154771-1eba5980-d2db-11e8-8b10-184ee6e52d43.png">
however it appears that Chrome and VoiceOver don't read out aria-live statuses. 

If you check this testing page https://terrillthompson.com/tests/aria/live-scores.html, Safari reads everything out as expected but Chrome does absolutely nothing. The same is true in this example by W3C https://www.w3.org/TR/WCAG20-TECHS/ARIA19.html . I'm not sure if we will be able to get around this issue.

owencjones commented:
> If you check this testing page https://terrillthompson.com/tests/aria/live-scores.html, Safari reads everything out as expected but Chrome does absolutely nothing. The same is true in this example by W3C https://www.w3.org/TR/WCAG20-TECHS/ARIA19.html . I'm not sure if we will be able to get around this issue.

At that point, I think you can safely say that it's a bug with the browser, rather than your code.

Maybe there's a way to come up with a workaround, but it's very possibly not.  Comes down to whether it's more important to have typeahead without arialive, or not have it, which is a design decision.

Could possibly find a feature detection method to work out whether we can detect whether they will be read, and if they can't, just switch to a more fallback way of doing things?

owencjones commented on .gitignore:
```diff
@@ -5,5 +5,7 @@ public
 yarn-error.log
 coverage
+env
+.vscode
```
👍 If anything, I'd include a few more, but obviously that's out of scope for this, so no need to.

I _think_ we could safely ignore anything matching `*.log` or beginning with a dot - I don't know of a time we need to commit dotfiles.

bameyrick replied:
I agree, I will add a few more... I've been getting .tmp files knocking around too which should really be ignored.

owencjones commented on components/02-form-elements/typeahead/README.md:
```diff
@@ -0,0 +1,57 @@
+
+When the JavaScript loads it will hide the select and show a text input. It will also change the for attribute on the label to match the id of the text input. When the user selects a value it will update the value of the select and fire an onchange event on the select. If the user types an exact match for a value (case insensitive) it will automatically select this value.
+
+Alternatives can by added as a coma separated list. When a user selects an option via an alternative the matched alternative will be shown in brackets next to the suggestion, however the main text and value will still be used (i.e. User searches _Wales_ - Suggestion will be _United Kingdom (Wales)_ - Selection will change input value to _United Kingdom_ and select value to _GB_).
```
```suggestion
Alternatives can by added as a comma separated list. When a user selects an option via an alternative the matched alternative will be shown in brackets next to the suggestion, however the main text and value will still be used (i.e. User searches _Wales_ - Suggestion will be _United Kingdom (Wales)_ - Selection will change input value to _United Kingdom_ and select value to _GB_).
```

A _coma_ separated list would be a very different thing!

bameyrick replied:
Haha i didn't even notice! But yes I expect everyone to enter some form of mild coma whilst separating the list 😉. 
(Will update now)

owencjones commented:
Have approved this, but obviously, subject to @armstrongb's approval too.

bameyrick commented:
@boxadesign hmm that used to be there, I must have broken that fixing some other issues. Will fix now

bameyrick commented:
@boxadesign  I slightly misunderstood your comment yesterday which is why my reply was nonsense, but I have fixed selecting options by clicking on them, and I've also added a "Continue typing to refine results" message to the bottom of the result list if results are limited and there are more results than the limit.
@will-madebymoon could you check whether you're happy with the addition message at the bottom of the results here please: https://deploy-preview-186--sdc-global-design-patterns.netlify.com/components/detail/typeahead--occupation-typeahead.html

armstrongb commented on components/02-form-elements/typeahead/README.md:
```diff
@@ -0,0 +1,56 @@
+When the JavaScript loads it will hide the select and show a text input. It will also change the for attribute on the label to match the id of the text input. When the user selects a value it will update the value of the select and fire an onchange event on the select. If the user types an exact match for a value (case insensitive) it will automatically select this value.
+
+Alternatives can by added as a comma separated list. When a user selects an option via an alternative the matched alternative will be shown in brackets next to the suggestion, however the main text and value will still be used (i.e. User searches _Wales_ - Suggestion will be _United Kingdom (Wales)_ - Selection will change input value to _United Kingdom_ and select value to _GB_).
+
```
I think we should note that this pattern may require the ability to link to a form allowing the user to search a data set or enter an address / value manually
```suggestion
**Note:** variants of this pattern should:
- include a link to a form allowing the user to add an address 'manually'
- to enter a value without selecting from the dataset
- to search a dataset and select a value from the returned results (no-js for larger datasets)
```

/cc @charlotte-hirst

bameyrick replied:
I've merged in my address picker branch which has a prototype for an address lookahead (under address rather than typeaheads) which should satisfy the user being able to enter the address manually.

I will create an example where the user does not necessarily have to choose from a predefined list next week.

armstrongb commented:
@bameyrick if you are happy that all issues have been resolved I will put this forward for re-testing with DAC.

armstrongb commented:
### Country typeahead - Input visual defect
![typeahead-overhang](https://user-images.githubusercontent.com/22910336/47650922-a32d9780-db79-11e8-91fd-4f2df20393cb.png)

When a result is longer than the input box there is a gap introduced.

bameyrick commented:
@armstrongb The overflow should be fixed now

saedwards commented:
Can we include unit tests for this work? Try to increase cross browser compatibility.

bameyrick commented:
@saedwards  I plan to write tests once it passes DAC testing. Didn't want to do them yet incase I have to do some big changes as a result of the DAC testing.

saedwards commented:
> @saedwards I plan to write tests once it passes DAC testing. Didn't want to do them yet incase I have to do some big changes as a result of the DAC testing.

Good stuff

armstrongb commented:
@bameyrick is this ready to merge or are there still unit tests to go in?
